common/cuda_hip/solver/batch_cg_kernels.hpp.inc

```diff
@@ -135,13 +137,45 @@ __global__ void apply_kernel(const int max_iter,
     auto warp_grp = group::tiled_partition<tile_size>(thread_block);
 
     for (size_type ibatch = blockIdx.x; ibatch < nbatch; ibatch += gridDim.x) {
+        const int gmem_offset =
+            ibatch * sconf.gmem_stride_bytes / sizeof(ValueType);
         extern __shared__ char local_mem_sh[];
-        ValueType* const r_sh = reinterpret_cast<ValueType*>(local_mem_sh);
-        ValueType* const z_sh = r_sh + nrows;
-        ValueType* const p_sh = z_sh + nrows;
-        ValueType* const Ap_sh = p_sh + nrows;
-        ValueType* const x_sh = Ap_sh + nrows;
-        ValueType* const prec_work_sh = x_sh + nrows;
+        ValueType* r_sh;
+        ValueType* z_sh;
+        ValueType* p_sh;
+        ValueType* Ap_sh;
+        ValueType* x_sh;
+        ValueType* prec_work_sh;
+        if (sconf.n_shared >= 1) {
+            r_sh = reinterpret_cast<ValueType*>(local_mem_sh);
+        } else {
+            r_sh = workspace + gmem_offset;
+        }
+        if (sconf.n_shared == 1) {
+            z_sh = workspace + gmem_offset;
+        } else {
+            z_sh = r_sh + sconf.padded_vec_len;
+        }
+        if (sconf.n_shared == 2) {
+            p_sh = workspace + gmem_offset;
+        } else {
+            p_sh = z_sh + sconf.padded_vec_len;
+        }
+        if (sconf.n_shared == 3) {
+            Ap_sh = workspace + gmem_offset;
+        } else {
+            Ap_sh = p_sh + sconf.padded_vec_len;
+        }
+        if (!sconf.prec_shared && sconf.n_shared == 4) {
+            prec_work_sh = workspace + gmem_offset;
+        } else {
+            prec_work_sh = Ap_sh + sconf.padded_vec_len;
+        }
+        if (sconf.n_shared == 5) {
```

Comment:
Yes, see line 169. This if loop is for the x vector.

Reply:
Think about what happens if `n_shared == 4` and `prec_shared == true`. Then, the preconditioner is in shared memory, while we need `x_sh` to go to global memory. As it stands, this will take the `else` branch, and `x_sh` will be in limbo. This only makes sense if the preconditioner were also regarded as a vector, but it's not.